Support Multiple inverter api for inverter type

kdehairy opened the pull request:
The main problem is that inverters apis are different in two ways:

- The api end point to get the data
- Some aspects of parsing the non-"data" attributes of the response (like SN, version, etc...)

Each single inverter could expose (for example) two end points in general. as in, before version x it exposes an end point with query parameters, but after version x it exposes the end point with params in the request body.

The inheritance model in the project is short handed here. As it is based on the assumption that a single inverter will always support a single api end point contract.

This patch is basically (and only for now), refactoring in favor of composition (vs. inheritance).

Each inverter class is initiated with two dependencies/components:

- InverterHttpClient
- ResponseParser

To support multiple variants for each inverter, we only build the inverter with proper variations of those dependencies.

So, After this patch, the discovery process is all about asking the inverter class to build all possible variants this inverter knows it supports. The discovery then goes on, as was before, trying to use that variant to get_data, until one variant is successful.

The patch tries as much as possible to continue providing sane and most common defaults for inveter's implementations.

At this moment, the patch does not add (or subtract) any feature/support on top of the baseline (before the patch). This is just opening the opportunity to easily add missing variants.

squishykid commented:
Thank you for the PR! I'll take some time to review it. In the mean time, it should now be possible for you to run the CI tooling.

My first comment is that there seem to be a lot of whitespace changes, which `black` is also complaining about. I'll swing back once the build is green!

kdehairy commented:
thnx. I amended my commit with the expected formatting. But I'm still not allowed to run the CI checks. Can you please give it a go for me?

squishykid commented:
I think I will have to approve each run until you get your first contribution in to the repo. After that it should run automatically

kdehairy commented:
Alright, I ran all the checks locally and fixed (to my best at least). Hopefully this next run will pass. Can you give it a go again? thnx

squishykid commented:
You should run the `black` code formatter locally before pushing, that will also reduce the size of the diff

kdehairy commented:
I did. But then I modified code to address the reset of the checks, and skipped my mind to run it again :) .. sorry

I fixed. And added a `verify.sh` script to run the checks locally. Might help next "first-contributor" to avoid back and forth communication with maintainers to run the checks manually for them.

squishykid commented:
I've been thinking about adding something like that. Thanks!

kdehairy commented:
Did you have a chance to review it? :)

squishykid commented on solax/__init__.py:
```diff
@@ -1,6 +1,5 @@
 """Support for Solax inverter via local API."""
 import asyncio
-
```
From an initial glance, I can see a few things we will need to change:
- There are a lot of whitespace and non-functional changes. I am referring to changes where you have reordered imports or removed whitespace between imports. These pollute the diff an will have to be removed
- I notice that you added a few logger statements. As this is a library, we should not be logging anything and leave the decision on how to bubble-up errors and information to the program calling the library, so you will have to remove the loggers too.

squishykid commented on solax/discovery.py:
```diff
@@ -26,6 +28,9 @@
     X1HybridGen4,
 ]
 
+_LOGGER = logging.getLogger(__name__)
```
These are one instance of the loggers that will have to be removed

squishykid commented on tests/test_solax.py:
```diff
@@ -1,6 +1,8 @@
-from unittest.mock import Mock
 import asyncio
```
It seems like there are no functional changes here, just reordering the imports. As mentioned above, this pollutes the diff.

squishykid commented on tests/test_vol.py:
```diff
@@ -1,5 +1,6 @@
 import pytest
 from voluptuous import Invalid
+
```
Whitespace change that pollutes the diff

squishykid commented on verify.sh:
```diff
@@ -0,0 +1,23 @@
+pylint -d 'C0111' solax tests
+
+echo "Running pytest..."
+pytest --cov=solax --cov-fail-under=100 --cov-branch --cov-report=term-missing .
```
Thank you for adding this file!

kdehairy commented:
Thanks for putting the time. I addressed the comments.

Although, I did not find a unified import sorting/ordering style in the code base (or any tool checking it). 

Do you find it a good idea to include something like `isort`?

squishykid commented on solax/inverters/__init__.py:
```diff
@@ -1,13 +1,13 @@
 from .x1_smart import X1Smart
 from .x3_v34 import X3V34
 from .x3 import X3
-from .x1_boost import X1Boost
```
There’s still an import re-shuffle here which is unnecessary

kdehairy commented:
Sorry I missed that. Addressed

kdehairy commented:
Any chance to have a look?

kdehairy commented:
Still no luck finding time for this PR?

squishykid commented:
Ok, lets do it!

squishykid commented:
Ok, i will need to have another quick check tomorrow, but I think we can merge this. Thank youuuuuu for your patience

kdehairy commented:
Thnx and non worries.

squishykid commented:
So my only issue here is how the inverters are constructed for the smoke tests, because it creates a list of them and then doesn’t exactly guarantee that it will test all of them.

I can fix that in a follow up PR though, so I’ll merge this and then pursue that in my own time. Thank you for this contribution, I really like the design you went with. And thanks for your patience!

kdehairy commented:
Thank you for putting the time reviewing.
